### PR TITLE
DRILL-4826: Query against INFORMATION_SCHEMA.TABLES degrades as the n…

### PR DESCRIPTION
**Pull request by parthchandra**
…umber of views increases

Changed to get information for all views in a single call instead of of one by one

### CODE REVIEW

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java [hunk 1]
```diff
@@ -78,17 +79,34 @@ public String getTypeName() {
...
-      return tableNameToTable;
+    final List<org.apache.hadoop.hive.metastore.api.Table> tables = Lists.newArrayList();
+
+    // In each round, Drill asks for a sub-list of all the requested tables
+    for(int fromIndex = 0; fromIndex < totalTables; fromIndex += bulkSize) {
+      final int toIndex = Math.min(fromIndex + bulkSize, totalTables);
+      final List<String> eachBulkofTableNames = tableNames.subList(fromIndex, toIndex);
+      List<org.apache.hadoop.hive.metastore.api.Table> eachBulkofTables;
+      // Retries once if the first call to fetch the metadata fails
+      synchronized(mClient) {
```
Comment (sudheeshkatkam) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: why synchronized?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 2]
```diff
@@ -738,5 +740,49 @@ public void dropTable(String table) {
             .build(logger);
       }
     }
+
+    @Override public List<Pair<String, TableType>> getTableNamesAndTypes(boolean bulkLoad, int bulkSize) {
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Add annotation in a line above?
There are other places too.

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java [hunk 3]
```diff
@@ -78,17 +79,34 @@ public String getTypeName() {
...
+        try {
+          eachBulkofTables = mClient.getTableObjectsByName(schemaName, eachBulkofTableNames);
```
Comment (sudheeshkatkam) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: + Why not use the helper? Exception handling and reconnecting logic is different in the helper methods in [DrillHiveMetaStoreClient](https://github.com/apache/drill/blob/master/contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/DrillHiveMetaStoreClient.java#L222). 
+ Move this logic to a method in that class?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java [hunk 4]
```diff
@@ -55,6 +54,7 @@
  * schema, table or field.
  */
 public abstract class InfoSchemaRecordGenerator {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(InfoSchemaRecordGenerator.class);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java]: private

File: exec/jdbc/src/test/java/org/apache/drill/jdbc/test/TestJdbcQuery.java [hunk 5]
```diff
@@ -122,6 +122,7 @@ public void testLikeNotLike() throws Exception{
       );
   }
 
+  @Ignore("Returns results in different order depeding on forkCount")
```
Comment (sudheeshkatkam) [inline exec/jdbc/src/test/java/org/apache/drill/jdbc/test/TestJdbcQuery.java]: Is this a regression due to this patch? Other wise, open a ticket for this issue.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java [hunk 6]
```diff
@@ -290,28 +290,30 @@ public Tables(OptionManager optionManager) {
...
         }
       }
     }
 
+    public boolean visitTableWithType(String schemaName, String tableName, TableType type) {
+      Preconditions
+          .checkNotNull(type, "Error. Type information for table %s.%s provided is null.", schemaName,
+              tableName);
+      records.add(new Records.Table(IS_CATALOG_NAME, schemaName, tableName, type.toString()));
+      return false;
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java]: why return `false`?

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java (hunk 1 above)
Comment (gparai) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: why do we synchronize?

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java [hunk 7]
```diff
@@ -78,17 +79,34 @@ public String getTypeName() {
...
-    List<org.apache.hadoop.hive.metastore.api.Table> tables;
-    try {
-      tables = DrillHiveMetaStoreClient.getTableObjectsByNameHelper(mClient, schemaName, tableNames);
-    } catch (TException e) {
-      logger.warn("Exception occurred while trying to list tables by names from {}: {}", schemaName, e.getCause());
-      return tableNameToTable;
+    final List<org.apache.hadoop.hive.metastore.api.Table> tables = Lists.newArrayList();
+
+    // In each round, Drill asks for a sub-list of all the requested tables
+    for(int fromIndex = 0; fromIndex < totalTables; fromIndex += bulkSize) {
```
Comment (gparai) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: Space?

File: exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java [hunk 8]
```diff
@@ -291,6 +291,9 @@
   String ENABLE_BULK_LOAD_TABLE_LIST_KEY = "exec.enable_bulk_load_table_list";
   BooleanValidator ENABLE_BULK_LOAD_TABLE_LIST = new BooleanValidator(ENABLE_BULK_LOAD_TABLE_LIST_KEY, false);
 
+  String BULK_LOAD_TABLE_LIST_BULK_SIZE_KEY = "exec.bulk_load_table_list.bulk_size";
```
Comment (gparai) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java]: Maybe a comment to describe the option?

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java (hunk 1 above)
Comment (parthchandra) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: This is refactored code from the fix for DRILL-4577. (https://github.com/apache/drill/pull/461)
I didn't really change it.
Going thru the code, it appears that m_client may be cached and reused and so probably should be synchronized.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java (hunk 2 above)
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: IntelliJ keeps reformatting this to be on the same line ! Will fix.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java (hunk 6 above)
Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java]: <sigh/> to keep it similar to visitTable which does the same, unnecessarily. I suppose I could change it to return void.

File: exec/jdbc/src/test/java/org/apache/drill/jdbc/test/TestJdbcQuery.java (hunk 5 above)
Comment (parthchandra) [inline exec/jdbc/src/test/java/org/apache/drill/jdbc/test/TestJdbcQuery.java]: Or maybe I should just order the results?

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java (hunk 1 above)
Comment (parthchandra) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: see previous comment

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java (hunk 7 above)
Comment (parthchandra) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: Where?

**Comment by parthchandra** [comment]
Updated to address review comments

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java [hunk 9]
```diff
@@ -231,4 +231,21 @@ public void dropTable(String tableName) {
...
   }
-}
\ No newline at end of file
+
+  public List<Pair<String, Schema.TableType>> getTableNamesAndTypes(boolean bulkLoad, int bulkSize) {
+    final List<String> tableNames = Lists.newArrayList(getTableNames());
+    final List<Pair<String, Schema.TableType>> tableNamesAndTypes = Lists.newArrayList();
+    final List<Pair<String, ? extends Table>> tables;
+    if (bulkLoad) {
+      tables = getTablesByNamesByBulkLoad(tableNames, bulkSize);
```
Comment (ppadma) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java]: why do we even have this option to do bulkLoad or not ? why not just do bulkLoad always ?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java [hunk 10]
```diff
@@ -290,28 +290,30 @@ public Tables(OptionManager optionManager) {
       return new PojoRecordReader<>(Records.Table.class, records.iterator());
     }
 
-    @Override
-    public void visitTables(String schemaPath, SchemaPlus schema) {
+    @Override public void visitTables(String schemaPath, SchemaPlus schema) {
```
Comment (ppadma) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/ischema/InfoSchemaRecordGenerator.java]: why this change ?

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java [hunk 11]
```diff
@@ -78,32 +79,49 @@ public String getTypeName() {
...
+        }
+        tables.addAll(eachBulkofTables);
+      }
     }
 
-    for(final org.apache.hadoop.hive.metastore.api.Table table : tables) {
-      if(table == null) {
+    final List<Pair<String, ? extends Table>> tableNameToTable = Lists.newArrayList();
+    for (final org.apache.hadoop.hive.metastore.api.Table table : tables) {
+      if (table == null) {
```
Comment (ppadma) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/schema/HiveDatabaseSchema.java]: can this table be null ?

File: contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/DrillHiveMetaStoreClient.java [hunk 12]
```diff
@@ -233,6 +236,30 @@ private DrillHiveMetaStoreClient(final HiveConf hiveConf) throws MetaException {
...
+    final List<org.apache.hadoop.hive.metastore.api.Table> tables = Lists.newArrayList();
+
+    // In each round, Drill asks for a sub-list of all the requested tables
+    for (int fromIndex = 0; fromIndex < totalTables; fromIndex += bulkSize) {
+      final int toIndex = Math.min(fromIndex + bulkSize, totalTables);
+      final List<String> eachBulkofTableNames = tableNames.subList(fromIndex, toIndex);
+      List<org.apache.hadoop.hive.metastore.api.Table> eachBulkofTables;
+      // Retries once if the first call to fetch the metadata fails
+      try {
+        eachBulkofTables =
```
Comment (sudheeshkatkam) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/DrillHiveMetaStoreClient.java]: `eachBulkofTables = getTableObjectsByNameHelper(mClient, schemaName, eachBulkofTableNames);`

Comment (parthchandra) [inline contrib/storage-hive/core/src/main/java/org/apache/drill/exec/store/hive/DrillHiveMetaStoreClient.java]: This is how intellij formats the code using the template we have.
